examples/common/pigweed/rpc_services/Attributes.h

```diff
@@ -173,24 +176,98 @@ class Attributes : public pw_rpc::nanopb::Attributes::Service<Attributes>
+        app::AttributeReportIBs::Builder attributeReports;
+        TLV::TLVWriter writer;
+        TLV::TLVType outer;
+        DeviceLayer::StackLock lock;
+
+        writer.Init(tlvBuffer);
+        PW_TRY(ChipErrorToPwStatus(writer.StartContainer(TLV::AnonymousTag(), TLV::kTLVType_Structure, outer)));
+        PW_TRY(ChipErrorToPwStatus(attributeReports.Init(&writer, kReportContextTag)));
+        PW_TRY(ChipErrorToPwStatus(app::ReadSingleClusterData(subjectDescriptor, false, path, attributeReports, nullptr)));
+        attributeReports.EndOfContainer();
+        PW_TRY(ChipErrorToPwStatus(writer.EndContainer(outer)));
+        PW_TRY(ChipErrorToPwStatus(writer.Finalize()));
+        tlvBuffer.reduce_size(writer.GetLengthWritten());
+
+        return ::pw::OkStatus();
+    }
+
+    template <typename T>
+    ::pw::Status TlvBufferGetData(ByteSpan tlvBuffer, TLV::TLVType expectedDataType, T & responseData)
+    {
+        TLV::TLVReader reader;
+        reader.Init(tlvBuffer);
+        // Open outer container
+        PW_TRY(NextAndCheckTlvTagAndType(&reader, TLV::AnonymousTag(), TLV::TLVType::kTLVType_Structure));
+        TLV::TLVType readerRoot;
+        PW_TRY(ChipErrorToPwStatus(reader.EnterContainer(readerRoot)));
+
+        // Open report container
+        PW_TRY(NextAndCheckTlvTagAndType(&reader, TLV::ContextTag(kReportContextTag), TLV::TLVType::kTLVType_Array));
+        TLV::TLVType readerArray;
+        PW_TRY(ChipErrorToPwStatus(reader.EnterContainer(readerArray)));
+
+        // Skip first array element which is the empty array from spec 10.5.4.3
+        PW_TRY(NextAndCheckTlvTagAndType(&reader, TLV::AnonymousTag(), TLV::TLVType::kTLVType_Structure));
+
+        // Parse the AttributeDataIB to pull out data
+        app::AttributeReportIB::Parser reportParser;
+        PW_TRY(ChipErrorToPwStatus(reportParser.Init(reader)));
+        app::AttributeDataIB::Parser dataParser;
+        PW_TRY(ChipErrorToPwStatus(reportParser.GetAttributeData(&dataParser)));
+        TLV::TLVReader dataReader;
+        PW_TRY(ChipErrorToPwStatus(dataParser.GetData(&dataReader)));
+        PW_TRY(CheckTlvTagAndType(&dataReader, TLV::ContextTag(0x2), expectedDataType));
+        PW_TRY(ChipErrorToPwStatus(dataReader.Get(responseData)));
+
+        return ::pw::OkStatus();
+    }
+
+    static ::pw::Status ChipErrorToPwStatus(CHIP_ERROR err)
+    {
+        if (err == CHIP_NO_ERROR)
+        {
+            return ::pw::OkStatus();
+        }
+        else if (err == CHIP_ERROR_BUFFER_TOO_SMALL)
+        {
+            return ::pw::Status::ResourceExhausted();
+        }
+        return ::pw::Status::Internal();
+    }
+
+    static ::pw::Status NextAndCheckTlvTagAndType(TLV::TLVReader * reader, TLV::Tag expectedTag, TLV::TLVType expectedType)
+    {
+        PW_TRY(ChipErrorToPwStatus(reader->Next()));
+        return CheckTlvTagAndType(reader, expectedTag, expectedType);
```

Comment:
Why isn't this just using the two-arg form of `Next()` instead of reinventing it?

Reply:
Good call, missed that. PR updated.